src/Sulu/Bundle/AdminBundle/Entity/CollaborationRepository.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\AdminBundle\Entity;
+
+use Psr\Cache\CacheItemPoolInterface;
+
+class CollaborationRepository
+{
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cache;
+
+    /**
+     * @var int
+     */
+    private $threshold;
+
+    public function __construct(CacheItemPoolInterface $cache, int $threshold)
+    {
+        $this->cache = $cache;
+        $this->threshold = $threshold;
+    }
+
+    public function find(string $resourceKey, string $id, string $connectionId): ?Collaboration
+    {
+        $cacheItem = $this->cache->getItem($this->getCacheId($resourceKey, $id));
+
+        if (!$cacheItem) {
+            return null;
+        }
+
+        $collaborations = $cacheItem->get();
+
+        if (!$collaborations) {
+            return null;
+        }
+
+        return $collaborations[$connectionId] ?? null;
+    }
+
+    /**
+     * @return Collaboration[]
+     */
+    public function update(Collaboration $collaboration): array
+    {
+        $collaboration->updateTime();
+
+        $cacheItem = $this->cache->getItem($this->getCacheIdFromCollaboration($collaboration));
+        $value = $cacheItem->get() ?? [];
+        $value[$collaboration->getConnectionId()] = $collaboration;
+
+        $value = array_filter($value, function(Collaboration $collaboration) {
+            return $collaboration->getChanged() > time() - $this->threshold;
+        });
+
+        $cacheItem->set($value);
+
+        $this->cache->save($cacheItem);
+
+        return array_values($value);
+    }
+
+    /**
+     * @return Collaboration[]
+     */
+    public function delete(Collaboration $collaboration): array
+    {
+        $cacheItem = $this->cache->getItem($this->getCacheIdFromCollaboration($collaboration));
+        $value = array_filter($cacheItem->get() ?? [], function(Collaboration $cachedCollaboration) use ($collaboration) {
+            return $collaboration->getConnectionId() !== $cachedCollaboration->getConnectionId();
+        });
+        $cacheItem->set($value);
+
+        $this->cache->save($cacheItem);
```

Comment:
remove cache entry, if $value is []

Reply:
The `delete` function is not always called, therefore it is not even sure that everything is deleted, if we do that here. For this reason we leave it as it is.